rfcs/rfc-story-based-git-workflow.md

```diff
@@ -0,0 +1,117 @@
+- Start Date: 2019-09-24
+- Target Major Version: N/A
+- Reference Issues: N/A
+- Entity Issue: 
+- Implementation PR: 
+
+# Summary
+
+This RFC proposes an approach to code management in the master branch of [https://github.com/bcgov/lear](https://github.com/bcgov/lear), specifically around branching, code reviews, and merging pull requests into the master branch.
+
+The proposal is to use a story branch workflow to increase the stability of the master branch by merging larger, more complete, and fully tested batches of work. The team agreement proposed is to shoot for a commit history on master that reflects changes story-by-story, where each story is complete and has been tested. 
+
+# Basic example
+
+Dev1 and Dev2 are working on a story that requires changes to the Vue application, the API, and the data model as well as end to end tests. 
+
+1. Dev1 and Dev2 have a conversation (probably including the PO or BA), reviewing the goal, the UX design, and technical approach.
+2. Dev1 and Dev2 create a story branch off master (choosing a name that identifies their story). 
+3. Working off a shared design and technical understanding, Dev1 begins the API and data model changes while Dev2 starts working up the screens. They can choose how they wish to commit to their shared branch (they can use pull requests and get each other to do code reviews if they want, do informal "over the shoulder" code reviews , or pair program).
+4. Dev1 and Dev2 work together to update the automated test coverage and ensure the tests all run in their local environment. They review their own code against common, published project coding standards and run any necessary static analysis tests, making corrections as necessary.
+5. As needed, Dev1 and Dev2 compare their branch with Master and rebase as necessary to keep from getting too far out of date. Rebasing early and often is easier than waiting, but it depends on the circumstaces. Dev1 and Dev2 stay in communication daily with the rest of the committers to stay abreast of changes that are coming soon to master.
+6. As needed, Dev1 and Dev2 compare their branch with other story branches that are currently in progress (same reason as above). 
+7. When Dev1 and Dev2 are satisfied that their story is complete and is up-to-date with MASTER, they create a pull request.
+8. The code is reviewed and a few small issues are corrected.
+9. The code is merged into master (squashing commits from the story branch) and the full test suite is re-run in the E2E environment. Because the tests have been run already, there is little chance of breaking the build at this point.
+10. The story branch is cleaned up (destroyed).
+
+# Motivation
+
+The original entities team ("A" team, "Coops" team, team "Aardvark") has been using an approach where they all fork the master branch individually and commit small changes as often as possible. The advantage to this approach is that it keeps merges incremental and small. 
```

Comment:
Close, we use one of the common GH patterns, sometimes referred to as  Forking Workflow. We do not work directly on Master.

Reply:
The current practice is to fork from master and PR into master as often as possible. That's exactly what is described here.